File: frontend/ckd-care-app/src/api/lab.ts
import { api } from "./client";

// 지표 정의 타입
export interface MetricDef {
  key: string;
  label: string;
  unit: string;
  decimals: number;
  range_low: number | null;
  range_high: number | null;
}

// 활성 지표 목록 응답
export interface MetricsResponse {
  active_keys: string[];
  active: MetricDef[];
  catalog: MetricDef[];
}

// 시계열 포인트
export interface LabPoint {
  date: string;
  value: number;
}

// 지표별 개요 (최근값·변화량·추세)
export interface MetricOverview {
  key: string;
  label: string;
  unit: string;
  decimals: number;
  latest: number | null;
  prev: number | null;
  delta: number | null;
  range_low: number | null;
  range_high: number | null;
  points: LabPoint[];
}

// 전체 개요 응답
export interface OverviewResponse {
  metrics: MetricOverview[];
  disclaimer: string;
}

// 저장 응답
export interface SaveLabResponse {
  measured_date: string;
  saved_keys: string[];
  auto_checkin: { performed: boolean; reason: string };
}

// 특정 날짜 기록 응답
export interface LabRecordResponse {
  measured_date: string | null;
  values: Record<string, number>;
  has_record: boolean;
}

// lab 검사 수치 API 클라이언트
export const labApi = {
  // 활성 지표 목록 조회
  getMetrics: () => api.get<MetricsResponse>("/records/lab/metrics"),

  // 활성 지표 목록 수정
  setMetrics: (metric_keys: string[]) =>
    api.put<MetricsResponse>("/records/lab/metrics", { metric_keys }),

  // 전체 지표 개요 조회 (최근값·추세·참고범위)
  getOverview: () => api.get<OverviewResponse>("/records/lab/overview"),

  // 특정 날짜 기록 조회
  getRecord: (date: string) =>
    api.get<LabRecordResponse>(`/records/lab?date=${date}`),

  // 검사 결과 저장 (날짜 + 지표값 맵)
  saveRecord: (measured_date: string, values: Record<string, number>) =>
    api.put<SaveLabResponse>("/records/lab", { measured_date, values }),

  // 특정 날짜 기록 삭제
  deleteRecord: (date: string) =>
    api.delete<LabRecordResponse>(`/records/lab?date=${date}`),
};

File: app/dtos/lab.py
from datetime import date

from pydantic import BaseModel

from app.dtos.base import BaseSerializerModel
from app.dtos.record import AutoCheckinResult


class MetricDef(BaseSerializerModel):
    """검사 지표 정의"""

    key: str
    label: str
    unit: str
    decimals: int
    range_low: float | None = None
    range_high: float | None = None


class MetricsResponse(BaseSerializerModel):
    """활성 지표 목록 조회 응답"""

    active_keys: list[str]
    active: list[MetricDef]
    catalog: list[MetricDef]


class SetMetricsRequest(BaseModel):
    """활성 지표 설정 요청"""

    metric_keys: list[str]


class SaveLabRequest(BaseModel):
    """검사 수치 저장 요청"""

    measured_date: date
    values: dict[str, float]


class LabPoint(BaseSerializerModel):
    """검사 수치 시계열 포인트"""

    date: date
    value: float


class MetricOverview(BaseSerializerModel):
    """지표 개요 (최신/이전값, 차이, 범위, 시계열)"""

    key: str
    label: str
    unit: str
    decimals: int
    latest: float | None
    prev: float | None
    delta: float | None
    range_low: float | None
    range_high: float | None
    points: list[LabPoint]


class OverviewResponse(BaseSerializerModel):
    """검사 수치 개요 조회 응답"""

    metrics: list[MetricOverview]
    disclaimer: str


class SaveLabResponse(BaseSerializerModel):
    """검사 수치 저장 응답"""

    measured_date: date
    saved_keys: list[str]
    auto_checkin: AutoCheckinResult


class LabRecordResponse(BaseSerializerModel):
    """검사 기록 조회 응답"""

    measured_date: date | None
    values: dict[str, float]
    has_record: bool
